### PR TITLE
Fix gallery audit trail does not show full version string

### LINKED ISSUE #3738: Gallery audit trail does not show full version string
For this package:
https://www.nuget.org/packages/NDesk.Options.Redpoint/0.2.1

Audit trail shows:
```json
      "Version": "0.2.1",
      "NormalizedVersion": "0.2.1",
```

The .nuspec of this package contains:
```xml
<version>0.2.1+git.unspecified</version>
```

Not sure if this is repro'ing with `dev` code or not. Saw it on PROD.

### PR DESCRIPTION
**Pull request by xavierdecoster**
Fixes #3738

### CODE REVIEW

File: src/NuGetGallery/Controllers/PackagesController.cs [hunk 1]
```diff
@@ -1225,7 +1225,7 @@ private ActionResult EditFailed(string id, EditPackageRequest formData, Package
             TempData["Message"] = String.Format(
                 CultureInfo.CurrentCulture, Strings.SuccessfullyUploadedPackage, package.PackageRegistration.Id, package.Version);
 
-            return RedirectToRoute(RouteName.DisplayPackage, new { package.PackageRegistration.Id, package.Version });
+            return RedirectToRoute(RouteName.DisplayPackage, new { package.PackageRegistration.Id, package.NormalizedVersion });
```
Comment (xavierdecoster) [inline src/NuGetGallery/Controllers/PackagesController.cs]: This cannot be a separate PR as it is affected by the other changes in this PR. However, this already fixes this part of the SemVer2 spec: https://github.com/NuGet/Home/wiki/SemVer2-support-for-nuget.org-%28server-side%29#filter-by-default-nugetorg

> In addition, we should already fix any version normalization issues on the gallery frontend to leverage the latest NuGet.Versioning APIs, to ensure any future SemVer v2.0.0 package versions are properly displayed, and **accessible through the web site using the properly normalized version string in route parameters**.

File: tests/NuGetGallery.Facts/Services/PackageServiceFacts.cs [hunk 2]
```diff
@@ -1724,7 +1724,7 @@ public class ThePublishPackageMethod
                 package.PackageRegistration.Packages.Add(
                     new Package
                         {
-                            Version = "2.0",
+                            Version = "2.0.0",
```
Comment (xavierdecoster) [inline tests/NuGetGallery.Facts/Services/PackageServiceFacts.cs]: Looks like this test was relying on the fact that we called into `NuGetVersion.ToString()`. Transitioning to `.ToFullString()` means this test was no longer properly arranged.

File: tests/NuGetGallery.Facts/Services/PackageServiceFacts.cs [hunk 3]
```diff
@@ -384,7 +384,7 @@ public void WillMakeTheCurrentUserTheOwnerWhenCreatingANewPackageRegistration()
                 // That said, it's still asserting one "thing": that the package data was read.
                 // I'm sorry, but I just can't imagine adding a test per property.
                 // Note that there is no assertion on package identifier, because that's at the package registration level (and covered in another test).
-                Assert.Equal("01.0.42.0", package.Version);
+                Assert.Equal("1.0.42", package.Version);
```
Comment (xavierdecoster) [inline tests/NuGetGallery.Facts/Services/PackageServiceFacts.cs]: This test was relying on the fact that the gallery called into `NuGetVersion.ToString()`. Transitioning to `.ToFullString()` means the test was no longer properly arranged.

File: tests/NuGetGallery.Facts/Services/PackageServiceFacts.cs [hunk 4]
```diff
@@ -384,7 +384,7 @@ public void WillMakeTheCurrentUserTheOwnerWhenCreatingANewPackageRegistration()
                 // That said, it's still asserting one "thing": that the package data was read.
                 // I'm sorry, but I just can't imagine adding a test per property.
                 // Note that there is no assertion on package identifier, because that's at the package registration level (and covered in another test).
```
Comment (scottbommarito) [inline tests/NuGetGallery.Facts/Services/PackageServiceFacts.cs]: nit: can we remove this passive-aggressive block of comments, excluding the last line?

Comment (xavierdecoster) [inline tests/NuGetGallery.Facts/Services/PackageServiceFacts.cs]: Lol! sure thing :)

**Comment by xavierdecoster** [comment]
@joelverhagen Correct. Having a property called `OriginalVersionString` could be handy :)

`ToString` does normalization for SemVer2 package versions instead of returning the original version string, whilst returning the non-modified original version string for non-SemVer2 packages.

`ToFullString` does normalization (e.g. stripping off leading zeros).

File: tests/NuGetGallery.Facts/Services/PackageServiceFacts.cs (hunk 4 above)
Comment (xavierdecoster) [inline tests/NuGetGallery.Facts/Services/PackageServiceFacts.cs]: Fixed with https://github.com/NuGet/NuGetGallery/pull/3801/commits/d1d6066785cdd40ded8be87e77fecbb4111fd58a

**Comment by joelverhagen** [comment]
> @joelverhagen Correct. Having a property called OriginalVersionString could be handy :)

We should make a PR on client side to get this. As it stands today, V2 and V3 will have different versions. feed2catalog gets the original value using XSLT:
https://github.com/NuGet/NuGet.Services.Metadata/blob/e407e41cf1bc20e67739a4d8717ccc6ce006b6bc/src/Catalog/xslt/nuspec.xslt#L131

**Comment by joelverhagen** [comment]
In my opinion, this is a change in behavior for V2. Previously, the Version property always had the original string. Now, SemVer 2.0.0 packages this is not the case. As mentioned before, this is also inconsistent with V3.

We could work around by loading the .nuspec as XDocument and getting the value ourselves.

**Comment by xavierdecoster** [comment]
Created client issue to track the API improvement on `NuGetVersion`: https://github.com/NuGet/Home/issues/5068

File: tests/NuGetGallery.Core.Facts/Packaging/ManifestValidatorFacts.cs [hunk 5]
```diff
@@ -477,7 +477,7 @@ public void ReturnsErrorIfDependencySetContainsInvalidId()
         {
             var nuspecStream = CreateNuspecStream(NuSpecDependencySetContainsInvalidId);
             
-            Assert.Equal(new[] { String.Format(Strings.ErrorInvalidPackageVersionForDependency, "jQuery", "packageA.1.0.1-alpha", string.Empty) }, GetErrors(nuspecStream));
+            Assert.Equal(new[] { $"Invalid package version for a dependency with id 'jQuery' in package 'packageA.1.0.1-alpha': ''" }, GetErrors(nuspecStream));
```
Comment (xavierdecoster) [inline tests/NuGetGallery.Core.Facts/Packaging/ManifestValidatorFacts.cs]: `Strings.ErrorInvalidPackageVersionForDependency` is no longer publicly accessible in `NuGet.Versioning`, so had to copy the string here.

**Comment by xavierdecoster** [comment]
@joelverhagen @scottbommarito updated PR to consume new `NuGetVersion.OriginalVersion` API. Can I have some more :eyes: on this one pls? :)

File: tests/NuGetGallery.Core.Facts/Packaging/ManifestValidatorFacts.cs (hunk 5 above)
Comment (joelverhagen) [inline tests/NuGetGallery.Core.Facts/Packaging/ManifestValidatorFacts.cs]: Could we capitalized ID?
https://github.com/NuGet/Home/issues/5018

Comment (xavierdecoster) [inline tests/NuGetGallery.Core.Facts/Packaging/ManifestValidatorFacts.cs]: Of course we can, but that first needs to be fixed in `NuGet.Versioning` :)

Comment (joelverhagen) [inline tests/NuGetGallery.Core.Facts/Packaging/ManifestValidatorFacts.cs]: Oh sorry, I misunderstood your comment. I thought this string was now mastered in NuGetGallery...

**Pull request merged**